File: Python/NextPermutation.py
class Solution:
    def nextPermutation(self, nums):
        n = len(nums)
        ind = -1

        # Step 1: Find the first decreasing element from the end
        for i in range(n - 2, -1, -1):
            if nums[i] < nums[i + 1]:
                ind = i
                break

        # Step 2: If not found, reverse the entire list
        if ind == -1:
            nums.reverse()
            return

        # Step 3: Find element just greater than nums[ind]
        for i in range(n - 1, ind, -1):
            if nums[i] > nums[ind]:
                nums[i], nums[ind] = nums[ind], nums[i]
                break

        # Step 4: Reverse the sublist from ind + 1 to end
        nums[ind + 1:] = reversed(nums[ind + 1:])

# Main function to test the solution
def main():
    sol = Solution()
    
    nums = [1, 2, 3]  # You can change this input to test other cases
    print("Original:", nums)
    
    sol.nextPermutation(nums)
    
    print("Next permutation:", nums)

if __name__ == "__main__":
    main()

File: Java/NextPermutation.java

import java.util.Arrays;


class Solution {
    public void nextPermutation(int[] nums) {
        int ind = -1;
        int n = nums.length;

        for(int i = n-2; i>=0;i--){
            if(nums[i]<nums[i+1]){
                ind = i;
                break;
            }
        }

        if(ind == -1){
           reverse(nums,0,n-1);
           return;
        }

        for(int i = n-1 ; i >= ind ; i--){
            if(nums[ind]< nums[i]){
                swap(nums,ind,i);
                break;
            }
        }
        reverse(nums,ind+1,n-1);
    }

    void swap(int []nums, int i,int j){
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

    void reverse(int [] nums, int start,int end){

        while(start<end){
            swap(nums,start,end);
            start++;
            end--;
        }
    }  
}


public class NextPermutation {
    public static void main(String[] args) {
        Solution solution = new Solution();

        int[] nums = {2, 1, 3};  // You can change this input to test other cases
        System.out.println("Original: " + Arrays.toString(nums));

        solution.nextPermutation(nums);

        System.out.println("Next permutation: " + Arrays.toString(nums));
    }
}